File: Chapter02/example_topic_publisher.rb
# Require client library
require 'bunny'

# Read RABBITMQ_URI from ENV
connection = Bunny.new ENV['RABBITMQ_URI']

# Start a communication session with RabbitMQ
connection.start
channel = connection.create_channel

# Ensure topic exists
def on_start(channel)
  # Declare and return the topic exchange, taxi-topic
  channel.topic("taxi-topic", durable: true, auto_delete: true)
end

# Publishing an order to the exchange
def order_taxi(type, exchange)
  payload = "example-message"
  message_id = exchange.publish(payload,
                                     routing_key: type,
                                     content_type: "application/json",
                                     content_encoding: "UTF-8",
                                     persistent: true,
                                     message_id: message_id)
end
exchange = on_start(channel)

#	Order will go to any eco taxi
order_taxi('taxi.eco', exchange)
#	Order will go to any taxi
order_taxi('taxi', exchange)

File: csharp_code/Chapter02/example_topic_publisher.cs
using System;
using System.Collections.Generic;
using System.Text;
using RabbitMQ.Client;
using RabbitMQ.Samples.Common;

namespace RabbitMQ.Samples.Chapter02
{
    public class ExampleTopicPublisher
    {
        public static void OrderTopicTaxis(RabbitMqConnection rabbitMqConnection)
        {
            var channel = rabbitMqConnection.Channel;
                
            OrderTaxi("taxi.eco.3", channel, "message for taxi.eco");
            OrderTaxi("taxi", channel, "message for taxi");
        }

        private static void OrderTaxi(string type, IModel channel, string payload)
        {
            // prepare the payload to be published
            var messageBodyBytes = Encoding.UTF8.GetBytes(payload);
            var props = channel.CreateBasicProperties();
            props.ContentType = "application/json";
            props.ContentEncoding = "UTF-8";
            props.Persistent = true;
            props.MessageId = Guid.NewGuid().ToString();

            // publish the message to the exchange
            channel.BasicPublish(exchange: Chapter2Constants.TopicExchange,
                routingKey: type,
                basicProperties: props,
                body: messageBodyBytes);
        }
    }
}
